### PR TITLE
GROOVY-7486: In groovy.util.Node.depthFirst please Provide a way to S…

### PR DESCRIPTION
**Pull request by paulk-asert**
…pecify pre-order, post-order Ordering

### CODE REVIEW

File: src/test/groovy/tree/NavigationTest.groovy [hunk 1]
```diff
@@ -57,11 +65,25 @@ class NavigationTest extends GroovyTestCase {
...
+
+    private static createTreeFromXmlParser() {
+        def xml = '''
+        <root>
+          <parent1>
+            <childa/>
+            <childb/>
+          </parent1>
+          <parent2>
+            <childa/>
```
Comment (keeganwitt) [inline src/test/groovy/tree/NavigationTest.groovy]: I think it'd improve this test to have all nodes uniquely named.

Comment (paulk-asert) [inline src/test/groovy/tree/NavigationTest.groovy]: Good idea, I'll make that change when I get a chance.

File: src/test/groovy/tree/NavigationTest.groovy [hunk 2]
```diff
@@ -22,29 +22,37 @@ package groovy.tree
...
+    }
+
+    void testPrePostOrder() {
+        def root = createTreeFromXmlParser()
+        def combos = [[false, true], ['depthFirst', 'breadthFirst']].combinations()
+        def actual = combos.collect{ preorder, type ->
+            root."$type"(preorder)*.name()
+        }*.toString().join('\n')
+        def expected = '''
+            [childa, childb, parent1, childa, grandchild, childb, childc, parent2, root]
```
Comment (keeganwitt) [inline src/test/groovy/tree/NavigationTest.groovy]: Might improve readability to use capital letters or numbers on children.

File: src/main/groovy/util/Node.java [hunk 3]
```diff
@@ -537,26 +539,40 @@ private NodeList getByName(String name) {
...
      */
     public List depthFirst() {
+        return depthFirst(true);
+    }
+
+    /**
+     * Provides a collection of all the nodes in the tree
+     * using a depth-first traversal.
+     *
+     * @param preorder if false, a postorder depth-first traversal will be performed
```
Comment (keeganwitt) [inline src/main/groovy/util/Node.java]: Solution doesn't provide In-order traversal, which was also requested in 7486.

File: src/test/groovy/tree/NavigationTest.groovy (hunk 2 above)
Comment (keeganwitt) [inline src/test/groovy/tree/NavigationTest.groovy]: Minor point, but might improve test readability by changing from a multiline string to a concatenated string with trailing comments labeling the combination.

File: src/main/groovy/util/Node.java (hunk 3 above)
Comment (paulk-asert) [inline src/main/groovy/util/Node.java]: As mentioned in the Jira, I am not aware of any widely used inorder traversal algorithms for non-binary trees. I have followed some of the computer science literature/research in that area but not all. Do you have references to such an algorithm?

File: src/test/groovy/tree/NavigationTest.groovy [hunk 4]
```diff
@@ -22,29 +22,37 @@ package groovy.tree
...
+            [root, parent1, childa, childb, parent2, childa, childb, grandchild, childc]
+            [grandchild, childa, childb, childc, childa, childb, parent1, parent2, root]
```
Comment (keeganwitt) [inline src/test/groovy/tree/NavigationTest.groovy]: I've never heard of pre/post order BFS.  Is that a thing?  I'd kinda have expected this method to reverse the order of each level (i.e. [root, parent2, parent1, childc, childb, childa, childb, childa, grandchild]).

File: src/main/groovy/util/Node.java (hunk 3 above)
Comment (keeganwitt) [inline src/main/groovy/util/Node.java]: That's true, but one could convert the tree to a BST before performing the traversal.  Pretty expensive, but could be done.

Comment (paulk-asert) [inline src/main/groovy/util/Node.java]: I am not sure I understand where you are going with the BST idea but I have tweaked the tests a little as per your earlier comments.

Comment (keeganwitt) [inline src/main/groovy/util/Node.java]: You'd traverse the tree (traversal type doesn't matter), adding each node to a TreeSet, then return a List of the TreeSet.  Kinda funky, I'm not sure it's a good idea.  It's just the only way I could think of to offer in-order traversal for non-binary trees.

**Comment by paulk-asert** [comment]
merged

**Pull request closed without merging**